flow/connectors/snowflake/information_schema_cache.go

```diff
@@ -0,0 +1,250 @@
+package connsnowflake
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"log/slog"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/PeerDB-io/peer-flow/connectors/utils"
+	"github.com/PeerDB-io/peer-flow/generated/protos"
+	"github.com/PeerDB-io/peer-flow/model/qvalue"
+	"github.com/PeerDB-io/peer-flow/peerdbenv"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+)
+
+const (
+	schemaExistsSQL = "SELECT TO_BOOLEAN(COUNT(1)) FROM INFORMATION_SCHEMA.SCHEMATA WHERE UPPER(SCHEMA_NAME)=?"
+	tableExistsSQL  = `SELECT TO_BOOLEAN(COUNT(1)) FROM INFORMATION_SCHEMA.TABLES WHERE UPPER(TABLE_SCHEMA)=? and UPPER(TABLE_NAME)=?`
+	tableSchemaSQL  = `SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE UPPER(TABLE_SCHEMA)=? AND UPPER(TABLE_NAME)=? ORDER BY ORDINAL_POSITION`
+	tableSchemasInSchema = `SELECT TABLE_NAME, COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE UPPER(TABLE_SCHEMA)=? ORDER BY TABLE_NAME, ORDINAL_POSITION`
+)
+
+type informationSchemaCache struct {
+	tableSchemaCache  *expirable.LRU[string, *protos.TableSchema]
+	schemaExistsCache *expirable.LRU[string, bool]
+	tableExistsCache  *expirable.LRU[string, bool]
+}
+
+func newInformationSchemaCache() *informationSchemaCache {
+	schemaExistsCache := expirable.NewLRU[string, bool](100_000, nil, time.Hour*1)
+	tableExistsCache := expirable.NewLRU[string, bool](100_000, nil, time.Hour*1)
+
+	tsCacheExpiry := peerdbenv.PeerDBSnowflakeTableSchemaCacheSeconds()
+	tableSchemaCache := expirable.NewLRU[string, *protos.TableSchema](100_000, nil, tsCacheExpiry)
+
+	return &informationSchemaCache{
+		tableSchemaCache:  tableSchemaCache,
+		schemaExistsCache: schemaExistsCache,
+		tableExistsCache:  tableExistsCache,
+	}
+}
+
+var (
+	cache     *informationSchemaCache
+	cacheInit sync.Once
+)
+
+type SnowflakeInformationSchemaCache struct {
+	ctx      context.Context
+	database *sql.DB
+	logger   slog.Logger
+}
+
+func NewSnowflakeInformationSchemaCache(
+	ctx context.Context,
+	database *sql.DB,
+	logger slog.Logger,
+) *SnowflakeInformationSchemaCache {
+	cacheInit.Do(func() {
+		cache = newInformationSchemaCache()
+	})
+	return &SnowflakeInformationSchemaCache{
+		ctx:      ctx,
+		database: database,
+		logger:   logger,
+	}
+}
+
+// SchemaExists returns true if the schema exists in the database.
+func (c *SnowflakeInformationSchemaCache) SchemaExists(schemaName string) (bool, error) {
+	schemaName = strings.ToUpper(schemaName)
+
+	if cachedExists, ok := cache.schemaExistsCache.Get(schemaName); ok {
```

Comment:
bit unfortunate that they made an LRU expirable cache without an atomic `GetOrUpsert` method